processing/src/main/java/io/druid/segment/incremental/IncrementalIndex.java

```diff
@@ -384,55 +411,127 @@ public int add(InputRow row) throws IndexSizeExceededException
+      final Boolean raceWinner;
+      if (dimensionOrder.containsKey(dimension)) {
+        // Common case, don't even try to lock
+        raceWinner = false;
+      }else{
+        final Lock dimLock = dimensionSynchronizer.get(dimension);
+        dimLock.lock();
+        try {
+          // If we didn't win the race, populate cache values in the same codepath below as if we never tried to lock
+          if(raceWinner = !dimensionOrder.containsKey(dimension)){
+            // we won a race, add a new dimension
+            // Set column capabilities as data is coming in if it wasn't set in constructor
+            ColumnCapabilitiesImpl capabilities = columnCapabilities.get(dimension);
+            if (capabilities == null) {
+              capabilities = new ColumnCapabilitiesImpl();
+              capabilities.setType(ValueType.STRING);
+              ColumnCapabilitiesImpl other = columnCapabilities.putIfAbsent(dimension, capabilities);
+              if (other != null) {
+                throw new ISE("Bad columnCapabilites state");
+              }
+            }
+            if (dimensionValues.size() > 1) {
+              capabilities.setHasMultipleValues(true);
+            }
+
+            final Integer orderID = orderIncrementer.getAndIncrement();
+            if (null == dims) {
+              dims = new String[dimSizer(rowDimensions)][];
+            } else if (orderID >= dims.length) {
+              dims = Arrays.copyOf(dims, dimSizer(rowDimensions));
+            }
+            dims[orderID] = canonizeDimValues(dimValues.addIfAbsent(dimension), dimensionValues);
+            dimensionOrder.put(dimension, orderID);
+          }
+        }
+        finally {
+          dimLock.unlock();
         }
-        if (dimensionValues.size() > 1) {
-          capabilities.setHasMultipleValues(true);
+      }
+      if (!raceWinner) {
+        final Integer orderID = dimensionOrder.get(dimension);
+        if (null == dims) {
+          dims = new String[dimSizer(rowDimensions)][];
+        } else if (orderID >= dims.length) {
+          dims = Arrays.copyOf(dims, dimSizer(rowDimensions));
         }
+        dims[orderID] = canonizeDimValues(dimValues.get(dimension), dimensionValues);
+      }
+    }
+
+    if (dims == null) {
+      // If this row has no dimensions for some reason
+      dims = new String[orderIncrementer.get()][];
+    }
 
-        Integer index = dimensionOrder.get(dimension);
-        if (index == null) {
-          dimensionOrder.put(dimension, dimensionOrder.size());
-          dimensions.add(dimension);
+    final TimeAndDims key = new TimeAndDims(Math.max(gran.truncate(row.getTimestampFromEpoch()), minTimestamp), dims);
+
+
+    // Now do the aggregation
+    Integer rowIndex;
 
-          if (overflow == null) {
-            overflow = Lists.newArrayList();
+    List<AggregatorType> aggs;
+
+    rowIndex = getFacts().putIfAbsent(key, RACE_CHECK_PLACEHOLDER);
+    if (rowIndex != null) {
```

Comment:
If you are doing an if/else it's usually a lot easier to read the code when the logic is not inverted.  I.e., check for `== null`, move the code from the else into the if and the if into the else.